code/segmentedtickerplant/stplog.q

```diff
@@ -0,0 +1,303 @@
+replaylog:{[t]
+  getlogs[replayperiod][t]
+ }
+
+// alternative replay allows for 'pass through logging'
+// if SCTP not producing logs, subscribers replay from STP log files
+if[.sctp.loggingmode=`parent;
+  replaylog:{[t]
+    .sctp.tph (`.stplg.replaylog; t)
+    }
+  ]
+
+getlogs:enlist[`]!enlist ()
+
+// If replayperiod set to `period, only replay logs for current logging period
+getlogs[`period]:{[t]
+  distinct flip (.stplg.msgcount;exec tbl!logname from `..currlog where tbl in t)@\:t
+ };
+
+// If replayperiod set to `day, replay all of today's logs
+getlogs[`day]:{[t]
+  // set the msgcount to 0Wj for all logs which have closed
+  lnames:select seq,tbls,logname,msgcount:0Wj from .stpm.metatable where any each tbls in\: t;
+  // Meta table does not store counts for live logs, so these are populated here
+  lnames:update msgcount:sum each .stplg.msgcount[tbls] from lnames where seq=.stplg.i;
+  flip value exec `long$msgcount,logname from lnames
+ };
+
+//////////////////////////////////////////////////////////////////////////////////////
+
+// Open log for a single table at start of logging period
+openlog:{[multilog;dir;tab;p]
+  lname:logname[multilog][dir;tab;p];
+  .lg.o[`openlog;"opening logfile: ",string lname];
+  h:$[(notexists:not type key lname)or null h0:exec first handle from `..currlog where logname=lname;
+    [if[notexists;.[lname;();:;()]];hopen lname];
+    h0
+  ];
+  `..currlog upsert (tab;lname;h);
+ };
+
+// Error log for failed updates in error mode
+openlogerr:{[dir]
+  lname:.[{hsym`$string[x],"/",string[y],(raze string"dv"$(.z.p+.eodtime.dailyadj)) except".:"};(dir;`$"_" sv string .proc.procname,errorlogname);{.lg.e[`openlogerr;"failed to make error log: ",x]}];
+  // lname:`$"_" sv string .proc.procname,lname;
+  if[not type key lname;.[lname;();:;()]];
+  h:@[{hopen x};lname;{.lg.e[`openlogerr;"failed to open handle to error log with error: ",x]}];
+  `..currlog upsert (errorlogname;lname;h);
+ };
+
+// Log failed message and error type in error mode
+badmsg:{[e;t;x]
+  .lg.o[`upd;"Bad message received, error: ",e];
+  `..loghandles[errorlogname] enlist(`upderr;t;x);
+ };
+
+closelog:{[tab]
+  if[null h:`..currlog[tab;`handle];.lg.o[`closelog;"no open handle to log file"];:()];
+  .lg.o[`closelog;"closing log file ",string `..currlog[tab;`logname]];
+  @[hclose;h;{.lg.e[`closelog;"handle already closed"]}];
+  update handle:0N from `..currlog where tbl=tab;
+ };
+
+// Roll all logs at end of logging period
+rolllog:{[multilog;dir;tabs;p]
+  .stpm.updmeta[multilog][`close;tabs;p];
+  closelog each tabs;
+  @[`.stplg.msgcount;tabs;:;0];
+  {[m;d;t]
+    .[openlog;(m;d;t;currperiod);
+      {.lg.e[`stp;"failed to open log for table ",string[y],": ",x]}[;t]]
+  }[multilog;dir;]each tabs;
+  .stpm.updmeta[multilog][`open;tabs;p];
+ };
+
+// creates dictionary of process data to be used at endofday/endofperiod
+endofdaydatadef:{`proctype`procname`tables!(.proc.proctype;.proc.procname;.stpps.t)};
```

Comment:
I don't think you need to define this separately - this definition can just be the catch part of the line below